model/generate.go

```diff
@@ -1,25 +1,296 @@
+		g.BuildVariants = append(g.BuildVariants, *bvs[i])
+	}
+	for i := range tasks {
+		g.Tasks = append(g.Tasks, *tasks[i])
+	}
+	g.Functions = functions
+	return g
 }
 
-// GeneratedTasksFromJSON returns a GeneratedTasks type from JSON. We use the
+// ParseProjectFromJSON returns a GeneratedTasks type from JSON. We use the
 // YAML parser instead of the JSON parser because the JSON parser will not
 // properly unmarshal into a struct with multiple fields as options, like the YAMLCommandSet.
-func GeneratedTasksFromJSON(data []byte) (*GeneratedTasks, error) {
-	g := &GeneratedTasks{}
-	if err := yaml.Unmarshal(data, g); err != nil {
-		return nil, errors.Wrap(err, "error unmarshaling into GeneratedTasks")
+func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
+	g := GeneratedProject{}
+	if err := yaml.Unmarshal(data, &g); err != nil {
+		return g, errors.Wrap(err, "error unmarshaling into GeneratedTasks")
 	}
 	return g, nil
 }
+
+// AddGeneratedProjectToVersion adds the buildvariants, tasks, and functions
+// from a generated project config to a project.
+func (g *GeneratedProject) AddGeneratedProjectToVersion() error {
+	// Get task, version, and project.
+	t, err := task.FindOneId(g.TaskID)
+	if err != nil {
+		return errors.Wrapf(err, "error finding task %s", g.TaskID)
+	}
+	v, err := version.FindOneId(t.Version)
+	if err != nil {
+		return errors.Wrapf(err, "error finding version %s", t.Version)
+	}
+	p := &Project{}
+	if err := LoadProjectInto([]byte(v.Config), t.Project, p); err != nil {
+		return errors.Wrap(err, "error reading project yaml")
+	}
+
+	// Cache project data in maps for quick lookup
+	cachedProject := cacheProjectData(p)
+
+	// Validate generated project against original project.
+	if err := g.validateGeneratedProject(p, cachedProject); err != nil {
+		return errors.Wrap(err, "generated project is invalid")
+	}
+
+	// Create and save new config. This is definitely racy.
+	config, err := g.addGeneratedProjectToConfig(v.Config, cachedProject)
+	if err != nil {
+		return errors.Wrap(err, "error creating config from generated config")
+	}
+	v.Config = config
+	version.UpdateOne(version.ById(v.Id), bson.M{"$set": bson.M{version.ConfigKey: config}})
+
+	// Save new builds and tasks to the db.
+	if err := LoadProjectInto([]byte(config), t.Project, p); err != nil {
+		return errors.Wrap(err, "error reading project yaml")
+	}
+	if err := g.saveNewBuildsAndTasks(cachedProject, v, p); err != nil {
+		return errors.Wrap(err, "error savings new builds and tasks")
+	}
+
+	return nil
+}
+
+func cacheProjectData(p *Project) projectMaps {
+	cachedProject := projectMaps{
+		buildVariants: map[string]struct{}{},
+		tasks:         map[string]*ProjectTask{},
+		functions:     map[string]*YAMLCommandSet{},
+	}
+	// use a set because we never need to look up buildvariants
+	for _, bv := range p.BuildVariants {
+		cachedProject.buildVariants[bv.Name] = struct{}{}
+	}
+	for _, t := range p.Tasks {
+		cachedProject.tasks[t.Name] = &t
```

Comment:
does this have to be a pointer? I think it's fine, but it gives me chills.

Reply:
it's potentially a lot of copying if not